### PR TITLE
[3.0] Improve tgz file handling

### LINKED ISSUE #7227: Package manager won't parse some packages generated on OSX
#### Description
When I use a package that was generated on OS X, I get these errors during upload:
```
( ! ) Fatal error: Uncaught TypeError: vsprintf(): Argument #2 ($values) must be of type array, string given in /Sources/Errors.php on line 231
( ! ) TypeError: vsprintf(): Argument #2 ($values) must be of type array, string given in /Sources/Errors.php on line 231
```
This doesn't seem to be the root of the error, but what I'm getting back

If I manually upload the file, nothing shows up.

The file was generated using OSX's built in tar uitlity
```
➜  StopForumSpam git:(master) tar -zcf ~/Downloads/StopForumSpam-test.tgz *
```

### Steps to reproduce
1. Attempt to upload the attached file to package manger.  Errors
2. 

### Environment (complete as necessary)
- Version/Git revision: latest
- Database Type: N/A
- Database Version: N/A
- PHP Version: 8.0

### Additional information/references
[StopForumSpam-test.tar.gz](https://github.com/SimpleMachines/SMF2.1/files/7835728/StopForumSpam-test.tar.gz)

OS X tar utility is adding some additional metadata (which is not a file) to the compressed file, which SMF Package Manager is trying to read out.

You should see data like this in the archive.
```
119 LIBARCHIVE.xattr.com.apple.quarantine=MDA4MTs2MWRhMDUxNTtDaHJvbWU7MEY5NERCOTgtQ0M0Ni00ODIwLTk5MEEtM0M0QTRBODY4RkVC 95 SCHILY.xattr.com.apple.quarantine=0081;61da0515;Chrome;0F94DB98-CC46-4820-990A-3C4A4A868FEB
```

This only happens I think if a file exists in the directory that is quarantined by OS X.  https://apple.stackexchange.com/questions/104712/what-causes-os-x-to-mark-a-folder-as-quarantined

### PR DESCRIPTION
**Pull request by live627**
- Fixes #7227

### CODE REVIEW

**Comment by jdarwood007** [comment]
Is this compatible with #7228 or an alternative fix?

**Comment by live627** [comment]
This is alternative fix

**Comment by sbulen** [comment]
@live627 - two things:
 - It doesn't fix the initial problem, I am still seeing the same/similar issues when the OSX files are present, e.g., WSOD in the Package Manager:
image

I am also seeing issues trying to analyze a package for the cust site - slightly different, though, it cannot find package-info.xml at all.

 - I am getting thousands of errors in the error log:
      2: Undefined array key "filename"   in Subs-Package, various lines, including 210, 206, 186, 178

I used the same test file we were using on the cust site with the OSX files.  (Since I don't have a mac I need to use @jdarwood007 's example...)

**Comment by sbulen** [comment]
I saw you pushed some commits while I was writing up my earlier test results.  

_I just retested, and all the issues reported above appear to have been addressed._

I must admit, though, that rewrites of stable code always make me nervous...  One of my most repeated complaints with PRs here...

**Comment by sbulen** [comment]
(When the destination is null, that is a request for a list of files in the archive.)

**Comment by jdarwood007** [comment]
If both fix the same thing, then I guess its a mater of choice.

This one feels like it may be more complete in trying to handle the tar file according to the RFC.

However we are late in 2.1 stages.  I see it also added hints.  While I like type hints, we have to be very careful of their usage in 2.1, because of so many places we could possibly break it.  This PR also may need additional testing for open_basedir restriction testing, which the `$destination !== null` checks fixed by preventing us from doing file checks on directories that would start from a path most likely outside of the allowed dirs.

In all honesty, I think #7228 should be merged for 2.1.  This one should go into the future because it is more complete into the RFC spec (as I see it handling some additional types other than files).  @Sesquipedalian may want to make that decision though.

**Comment by live627** [comment]
Testing this one is pretty much the same as https://github.com/SimpleMachines/SMF2.1/pull/7180#issuecomment-994229770

**Comment by sbulen** [comment]
@live627 - I'm not sure this is a hard requirement for 2.1.0?

**Comment by live627** [comment]
Not important enough to hold up the release. The excising code works fine with the ubiquitous ustar format, just not the "newer" pax format.

I'll  let @Sesquipedalian assign a new milestone as he sees fit.

**Comment by Sesquipedalian** [comment]
I'd rather merge the full fix. However, there's really no hurry for dealing with this issue. I've been making SMF compatible tar.gz files on my Mac for years; it's just a matter of setting the right command line switches. Anyway, we can take the time to test this PR as thoroughly as we like after 2.1.0 is out.

**Comment by jdarwood007** [comment]
FYI, this didn't get merged into 2.1.0.  Can we do the #7228 for 2.1.x?

**Comment by live627** [comment]
The speedup mentioned in the latest commit is 0.2 seconds with a package directory filled with 89 objects; package directories, php files, package files.

File: Sources/Packages.php [hunk 1]
```diff
@@ -1587,25 +1587,37 @@ function getPackages()
 
 	$installed_mods = loadInstalledPackages();
 
+	$packagesdir='C:\Users\John\nf\df';
```
Comment (jdarwood007) [inline Sources/Packages.php:1590]: Derp

File: Sources/Subs-Package.php [hunk 2]
```diff
@@ -146,56 +146,60 @@ function read_tgz_data(string $data, ?string $destination, bool $single_file = f
...
-				);
-
-				$checksum = 256;
-				for ($i = 0; $i < 512; $i++)
-					if (($i < 148 || $i > 155) && $data[($block << 9) + $i] != "\0")
-						$checksum += ord($data[($block << 9) + $i]);
+			if ($file_info['filename'][-1] !== '/')
+				$file_info['filename'] .= '/';
+		}
+		elseif ($file_info['type'] === 'L')
```
Comment (jdarwood007) [inline Sources/Subs-Package.php:154]: Since your updating these.  Can you document the types heres.  Still confusing.

**Comment by jdarwood007** [comment]
@live627 ,  Can you fix the conflict?

**Comment by Sesquipedalian** [comment]
Closed because the original issue has been resolved in 3.0 and I don't see any need to worry about it for 2.1.
